libs/sdk-core/src/ldk/store/locking_store.rs

```diff
@@ -0,0 +1,206 @@
+/// processes or instances might try to access the same storage concurrently.
+///
+/// The locking mechanism works by:
+/// 1. Acquiring a lock when the store is created
+/// 2. Refreshing the lock periodically to maintain ownership
+/// 3. Releasing the lock when explicitly unlocked
+///
+/// If another instance already holds the lock, attempting to create a new `LockingStore`
+/// will result in a `Conflict` error.
+pub struct LockingStore<S: VersionedStore + Send + Sync> {
+    inner: S,
+    tl: Mutex<TimeLock>,
+    version: AtomicI64,
+}
+
+#[allow(dead_code)]
+impl<S: VersionedStore + Send + Sync> LockingStore<S> {
+    const KEY: &str = "lock";
+    const LOCK_DURATION_SECS: u64 = 60;
+    const REFRESH_WINDOW_SECS: u64 = 30;
+
+    /// Creates a new `LockingStore` and attempts to acquire the distributed lock.
+    ///
+    /// This method will:
+    /// 1. Check if a lock already exists in the underlying store
+    /// 2. If no lock exists, create one for this instance
+    /// 3. If a lock exists, verify it's not held by another instance
+    /// 4. If the lock is held by another instance, return a `Conflict` error
+    pub async fn new(instance_id: String, store: S) -> Result<Self, Error> {
+        let (lock_data, version) = store.get(Self::KEY.to_string()).await?.unwrap_or_default();
+        let lock_duration = Duration::from_secs(Self::LOCK_DURATION_SECS);
+        let lock_data = LockData::decode(lock_data)
+            .map_err(|e| Error::Internal(format!("Failed to decode lock_data: {e:?}")))?;
+        let tl = TimeLock::new(lock_duration, instance_id, lock_data).map_err(
+            |LockedBy(instance_id)| {
+                Error::Conflict(format!("Remote lock aquired by `{instance_id}`"))
+            },
+        )?;
+        let tl = Mutex::new(tl);
+
+        let locking_store = Self {
+            inner: store,
+            tl,
+            version: AtomicI64::new(version),
+        };
+        locking_store.lock().await?;
+
+        Ok(locking_store)
+    }
+
+    /// Refreshes the distributed lock to extend its duration.
+    ///
+    /// This method should be called periodically to maintain ownership of the lock.
+    ///
+    /// # Returns
+    /// * `Ok(SystemTime)` - The time until which the lock should be refreshed
+    /// * `Err(Error::Conflict)` - If the lock was lost or cannot be refreshed
+    /// * `Err(Error::Internal)` - If there was an internal error
+    pub async fn refresh_lock(&self) -> Result<SystemTime, Error> {
+        let locked_until = self.lock().await?;
+        Ok(locked_until - Duration::from_secs(Self::REFRESH_WINDOW_SECS))
+    }
+
+    async fn lock(&self) -> Result<SystemTime, Error> {
+        let lock_data = self.tl.lock().await.next_lock();
+        let value = lock_data
+            .encode()
+            .map_err(|()| Error::Internal("Failed to encode lock_data".to_string()))?;
+        self.inner
```

Comment:
Should we also consider the case where  a lock exists with new version but it is expired? At this point shouldn't we fetch the new version and attempt to update from there?

Reply:
I see. IIUC the timelock will be created  when sdk connects. I am just worried if for example the app is going to the background without inet connection for some minutes/hours and meantime other instance acquired the lock and fail to release. In that case it seems we don't recover? This is not worrying you?